Add ability for a mobile::Module to save as flatbuffer

qihqi opened the pull request:
Included functions:

* save_mobile_module -> saves a mobile::Module to flatbuffer
* load_mobile_module_from_file -> loads a flatbuffer into mobile::Module
* parse_mobile_module -> parses from bytes or deserialized flatbuffer
      Module object

Fixes #{issue number}

qihqi commented on torch/csrc/jit/serialization/mobile_bytecode.fbs:
```diff
@@ -0,0 +1,178 @@
+  type_name:string;
+  type: TypeType;
+  // Below fields are optional
+  attr_names:[string];
```
attach function

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -0,0 +1,41 @@
+// returned mobile::Module and call deleter on mobile::Module's
+// destructor.
+TORCH_API mobile::Module parse_mobile_module(
+    void* data,
```
use unique_ptr to express ownership transfer

qihqi replied:
also use different name

qihqi replied:
parse_and_initialize_module

iseeyuan commented:
Just wondering what's the plan to land this PR? Would it be landed after the design review? Asking because the APIs look important and we may need more discussion on them.

qihqi commented:
> Just wondering what's the plan to land this PR? Would it be landed after the design review? Asking because the APIs look important and we may need more discussion on them.

My current plan is to land before design review then follow up with another PR fixing up any changes in API. From discussions from the doc it seems the compatibility APIs (the ones that is identical to the current ones) will need to wrap the new ones. But I am OK either way.

Feel free to start reviewing the implementation part as if there are concerns we can address them early.

zhxchen17 commented on torch/csrc/jit/serialization/mobile_bytecode.fbs:
```diff
@@ -0,0 +1,178 @@
+}
+
+table StorageData {
+  data: [ubyte] (force_align: 16);
```
I remember we don't need force_align anymore right? The reason is we enforce the alignment manually instead of relying on generated API from FB. Maybe put a comment here instead?

qihqi replied:
Turns out, (at least in the version of flatbuffer in BUCK, building for android) (force_align:16) doesnt do anything (i.e. could not find anything different from the generated headers). Now, we might still want to keep this annotation just to document intent.

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -0,0 +1,41 @@
+
+// Parse a mobile::Module from flatbuffer's in-memory Module representation.
+// The caller is assumed to manage the lifetimes of Module.
+TORCH_API mobile::Module parse_mobile_module(
```
initialize_mobile_module

iseeyuan reviewed (commented):
Went through part of this PR. Will continue from mobile folder.

pavithranrao commented on tools/build_variables.bzl:
```diff
@@ -485,6 +485,7 @@ torch_mobile_core = [
     # This should not be needed eventually.
     # TODO: Remove this dependency
     "torch/csrc/jit/backends/backend_debug_info.cpp",
+    "torch/csrc/jit/mobile/flatbuffer_loader.cpp",
```
added comments in Phabricator as it indirectly touches an internal file

cccclai replied:
I think it's used for OSS now. The change to use it as source of truth hasn't landed yet. 

Aside from it, if it's used by both jit and mobile, can it be added to `core_sources_common`, instead of adding it to both `torch_mobile_core` and `libtorch_extra_sources`?

iseeyuan replied:
@pavithranrao the internal one is not for production in general, but to be safe let's run the BSB to make sure not size regress is introduced. 

@cccclai It's a WIP PR that is not immediately to be used in production. It's estimated to be ~10s KB size increase in OSS runtime build. I assume it's fine because this increase would be there anyway for the transition period.

qihqi replied:
I removed those files from there; and moved to separate BUCK/bazel targets instead. This way we can choose to include them as deps when we roll it out.

iseeyuan commented on torch/csrc/jit/mobile/module.h:
```diff
@@ -130,12 +130,19 @@ class TORCH_API Module {
     return *cu_.get();
   }
 
+  void set_delete_memory(std::shared_ptr<char> delete_mem) {
```
What's the use case of this function? 

Sorry I may not fully follow the offline discussion on this.

qihqi replied:
It's used in `parse_and_initialize_mobile_module`. So we read some memory from the file; but cannot drop that memory yet because it is still used by the Module. So this allows us to attach the memory we got from file and to module so they have the same lifetime.

iseeyuan replied:
Thanks! Now I remember. It's for the memory part directly from the file, and to make it the same lifetime as the module.

iseeyuan commented on torch/csrc/jit/testing/module_differ.h:
```diff
@@ -0,0 +1,32 @@
+// 1. _ivalue() returned by both should be equal according to ivalueEquals below
+// 2. all functions with same name shall have same instructions and constants
+// 3. all functions in lhs exists in rhs.
+TORCH_API bool moduleEquals(
```
I assume it's used in serialization only, not for deserialization. Could you confirm?

qihqi replied:
this function is used for manual testing. It's not used in either serialization or loading

qihqi commented on tools/linter/clang_tidy/generate_build_files.py:
```diff
@@ -56,10 +56,19 @@ def run_autogen() -> None:
     )
 
 
+def gen_flatbuffers():
```
flatbuffer_loader.h / cpp etc includes the generated header. So without it it won't compile. clang_tidy compiles the file first then produce diagnostics. If compilation fails it makes a best effort in giving error messages which are mostly bogus.

iseeyuan reviewed (approved):
LGTM as the first version. More details can be added or adjusted when it's actually integrated to the mobile flow. 

Please remember to:
1. Add BSB result link in the internal diff for referrence.
2. Add the debug info in a followup PR.

qihqi committed:
Add ability for a mobile::Module to save as flatbuffer (#67351)

Summary:
Included functions:

* save_mobile_module -> saves a mobile::Module to flatbuffer
* load_mobile_module_from_file -> loads a flatbuffer into mobile::Module
* parse_mobile_module -> parses from bytes or deserialized flatbuffer
      Module object

Fixes #{issue number}

Pull Request resolved: https://github.com/pytorch/pytorch/pull/67351

Reviewed By: iseeyuan

Differential Revision: D32010095

Pulled By: qihqi

fbshipit-source-id: 11f43641033c9eaf96f9529a7925e781bba26821
